packages/firestore/test/integration/api/database.test.ts

```diff
@@ -1590,6 +1609,51 @@ apiDescribe('Database', (persistence: boolean) => {
         expect(docRef.isEqual(docRef2)).to.be.false;
       });
     });
+
+    it('Correct snapshot specified to fromFirestore() when registered with DocumentReference', () => {
+      return withTestDb(persistence, async db => {
+        const untypedDocRef = db.collection('/models').doc();
+        const docRef = untypedDocRef.withConverter(modelWithRefConverter);
+        await docRef.set(new ModelWithRef(docRef));
+        const docSnapshot = await docRef.get();
+        expect(untypedDocRef.isEqual(docSnapshot.data()!.ref)).to.be.true;
+      });
+    });
+
+    it('Correct snapshot specified to fromFirestore() when registered with CollectionReference', () => {
+      return withTestDb(persistence, async db => {
+        const untypedCollection = db
+          .collection('/models')
+          .doc()
+          .collection('sub');
+        const collection = untypedCollection.withConverter(
+          modelWithRefConverter
```

Comment:
I would expect that this converter does an instanceof check.

Reply:
I'm not sure what you are suggesting. Can you clarify?

Reply:
Ahh I see what you mean (from the Gist you posted: https://gist.github.com/schmidt-sebastian/7201fe874a6f463e0ad2256780386107). You were asking for an instanceof check in the fromFirestore() method. I've included this.